src/Network/Wai/Auth/Tools.hs

```diff
@@ -0,0 +1,46 @@
+{-# LANGUAGE BangPatterns #-}
+module Network.Wai.Auth.Tools
+  ( encodeKey
+  , decodeKey
+  , toLowerUnderscore
+  , getValidEmail
+  ) where
+
+import qualified Data.ByteString        as S
+import           Data.ByteString.Base64 as B64
+import           Data.Char              (isLower, toLower)
+import           Data.Foldable          (foldr')
+import           Data.Serialize         (Get, get, put, runGet, runPut)
+import qualified Data.Text              as T
+import           Web.ClientSession      (Key)
+
+
+-- | Decode a `Key` that is in a base64 encoded serialized form
+decodeKey :: Monad m => S.ByteString -> m Key
```

Comment:
I am not sure why you suggested `MonadThrow` here, there was no direct calling of `error` function or exception throwing, only a mere `fail`, which is still a part of `Monad`. In any case, in order to get rid of any possible partiality (due to `fail` implementation) I changed it to `Either` monad.